KAFKA-4527: task status was being updated before actual pause/resume

shikhar opened the pull request:
h/t @ewencp for pointing out the issue

shikhar committed:
KAFKA-4527: task status was being updated before actual pause/resume

h/t @ewencp for pointing out the issue

shikhar commented:
Still have to fix unit tests

ED: done

shikhar committed:
Status updates in WorkerTask are synchronized (see class javadoc)

shikhar committed:
Avoid paused-before-start status update, fix WorkerSourceTaskTest.testStartPaused()

shikhar committed:
Fix WorkerSinkTaskTest, add new expectations

shikhar committed:
Replace AtomicBoolean usage with volatile - no CAS happening

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -407,10 +407,12 @@ private void resumeAll() {
 
     private void pauseAll() {
         consumer.pause(consumer.assignment());
+        onPause();
```
Hmm, this one is kind of interesting. I was thinking pause would actually stop things, but I guess for sinks we can actually just pause the consumer. At first I thought we wouldn't want this because `pauseAll`/`resumeAll` were initially used just to pause the consumer when data was rejected with a `RetriableException`, which could, e.g., happen even if data is still being processed simply because a buffer got full or something.

But these methods are also used to just pause work when the target state is paused. This makes me wonder if we're also not doing quite enough in those cases before marking the task status as paused -- in fact the connector may still be processing data, we're just not feeding it anymore. I think we at least need a flush + commit, and to be really sure (e.g. in the case of the HDFS connector) we'd probably need to actually `close()` as well.

shikhar replied:
That's a good point. This could potentially be addressed by:
*  in `pauseAll()`, between pausing the consumer and the status callback, `closePartitions()` (which will take care of the commit also).
* in `resumeAll()`, as the very first thing call `openPartitions(consumer.assignment())`.

but there is interplay with the context and I'm not sure what the behavior should be. There may be some partitions that are currently paused by the sink task. The task can also pause/resume partitions using the context while we are paused. While paused, `context.assignment()` will continue to return the assigned partitions, despite us having called `task.close(partitions)` for them.

I think it's probably best to just pause record delivery to the task here.

shikhar replied:
Added a commit request via the context, so a commit will happen on the next iteration.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTask.java:
```diff
@@ -128,13 +125,10 @@ private void doClose() {
-                    statusListener.onPause(id);
-                else
-                    statusListener.onStartup(id);
+                statusListener.onStartup(id);
```
Hmm, we removed checking between `PAUSED` and `STARTED` target states here, but I don't see an equivalent check elsewhere. So it seems that, e.g., after a rebalance we will still create all the tasks and initialize them, but rely on `shouldPause()` checks to effectively put them to sleep. Is this the behavior we ultimately want or should paused tasks effectively not use any resources?

ewencp replied:
And importantly, I think this change will cause the status to change briefly after a rebalance -- giving the `onStartup` call will set the status to `RUNNING` and then the subsequent `onPause` will set it back to `PAUSED`.

shikhar replied:
>  So it seems that, e.g., after a rebalance we will still create all the tasks and initialize them, but rely on shouldPause() checks to effectively put them to sleep. Is this the behavior we ultimately want or should paused tasks effectively not use any resources?

I deliberately opted for this behavior in an attempt at keeping the flow straightforward, but it does have the drawback you mentioned. I hadn't thought about the oddness around rebalancing. I will make it so the task does not start until unpaused.

shikhar commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTask.java:
```diff
@@ -128,13 +125,10 @@ private void doClose() {
-                    statusListener.onPause(id);
-                else
-                    statusListener.onStartup(id);
+                statusListener.onStartup(id);
```
>  So it seems that, e.g., after a rebalance we will still create all the tasks and initialize them, but rely on shouldPause() checks to effectively put them to sleep. Is this the behavior we ultimately want or should paused tasks effectively not use any resources?

I deliberately opted for this in an attempt at keeping the flow straightforward, but it does have the drawback you mentioned. However I will make it so the task does not start at all until unpaused.

shikhar committed:
Don't start just to pause

shikhar commented:
The failing systest (see JIRA) passed here http://jenkins.confluent.io/job/system-test-kafka-branch-builder/633/

shikhar committed:
sink worker to request commit when pausing

shikhar committed:
checkstyle

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -407,10 +407,13 @@ private void resumeAll() {
 
     private void pauseAll() {
         consumer.pause(consumer.assignment());
+        context.requestCommit();
```
Is this ok to do for both reasons this would be called? Do we actually want to request a commit when in the case of the connector throwing a RetriableException? That'll result in the connector being forced to flush all its data.

shikhar replied:
Good catch, updated. I'd even updated the redelivery test to expect the commit but didn't realize that doesn't make much sense.

shikhar committed:
Update status callback and request commit only when pause/resume happens as a result of target state changing

The behavior when pausing/resuming due to RetriableException should be unchanged

ewencp commented:
LGTM, and tests are now passing: https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/170/console

Thanks @shikhar!
